### PR TITLE
removed unnecessary Fiber::ENDED

### PR DESCRIPTION
**Pull request by leeeqian**
so Thread.Count will be consistent with the live threads; also changed creating fiber to delete the fiber if it is ended, to handle "too many
fibers" problem when message loop does not get chance to clean up ended fibers.

### CODE REVIEW

File: fibers.cpp [hunk 1]
```diff
@@ -185,6 +184,8 @@ void Fibers::create(void (_stdcall *fiber_proc)(void* arg), void* arg)
 	for (int i = 1; i < MAXFIBERS; ++i)
 		if (fibers[i].status == Fiber::REUSE)
 			{
+			if (fibers[i].fiber != 0 && !fibers[i].fiber)
```
Comment (apmckinlay) [inline fibers.cpp]: Can this ever be true? Seems like it should just be ```if (fibers[fi].fiber)```

File: fibers.cpp [hunk 2]
```diff
@@ -241,14 +242,13 @@ bool Fibers::yield()
 	for (int i = 1; i < MAXFIBERS; ++i)
 		{
 		fi = fi % (MAXFIBERS - 1) + 1;
-		if (fibers[fi].status == Fiber::ENDED)
+		Fiber f = fibers[fi];
+		if (f.status == Fiber::REUSE && f.fiber != 0 && !f.fiber)
```
Comment (apmckinlay) [inline fibers.cpp]: See above comment

File: fibers.cpp [hunk 3]
```diff
@@ -241,14 +242,13 @@ bool Fibers::yield()
 	for (int i = 1; i < MAXFIBERS; ++i)
 		{
 		fi = fi % (MAXFIBERS - 1) + 1;
-		if (fibers[fi].status == Fiber::ENDED)
+		Fiber f = fibers[fi];
```
Comment (apmckinlay) [inline fibers.cpp]: If we make this Fiber& then we can do ```f = Fiber();``` below

File: fibers.cpp [hunk 4]
```diff
@@ -291,16 +291,16 @@ void Fibers::unblock(int fiberIndex)
...
+	cur->status = Fiber::REUSE;
 	yield();
 	unreachable();
 	}
 
 static void foreach_fiber(std::function<void(Fiber&)> f)
 	{
 	for (int i = 0; i < MAXFIBERS; ++i)
-		if (fibers[i].status < Fiber::ENDED)
+		if (fibers[i].status < Fiber::REUSE)
```
Comment (apmckinlay) [inline fibers.cpp]: Now that we don't have ENDED, we can eliminate '<' and just use ```!= Fiber::REUSE```

**Comment by leeeqian** [comment]
oops, I should test this change with logs, sorry.

I think I tried (fibers[fi].fiber) before, somehow I got access violation
problem, I will give it try again.

**Comment by leeeqian** [comment]
could this also be the reason why there are unlabeled connections, since the fiber could be never cleaned up properly?

File: fibers.cpp [hunk 5]
```diff
@@ -46,10 +46,9 @@
 #define LOG(stuff)
 
 static int fiber_count = 0;
-
```
Comment (apmckinlay) [inline fibers.cpp]: why remove this blank line?

File: fibers.cpp [hunk 6]
```diff
@@ -291,16 +296,16 @@ void Fibers::unblock(int fiberIndex)
 void Fibers::end()
 	{
 	LOG("end " << curFiberIndex());
-	verify(!inMain());
-	cur->status = Fiber::ENDED;
+	verify(!inMain());
```
Comment (apmckinlay) [inline fibers.cpp]: whitespace on end of line (might want to configure editor to remove this automatically)

File: fibers.cpp [hunk 7]
```diff
@@ -238,20 +242,21 @@ bool Fibers::yield()
...
+
+	for (int i = 1; i < MAXFIBERS; ++i)
+		{	
+		Fiber& f = fibers[i];
+		if (f.status == Fiber::REUSE && f.fiber)
 			{
-			switchto(fi);
-			return true;
+			LOG("delete fiber " << i << " from yield");
+			DeleteFiber(f.fiber);
```
Comment (apmckinlay) [inline fibers.cpp]: Although it shouldn't happen the way the code is, I think to be safe we should add an assert that we aren't deleting the current fiber.

File: fibers.cpp (hunk 5 above)
Comment (leeeqian) [inline fibers.cpp]: not intentionally, I run into a git conflict problem before, probably accidentally removed the line

I will put it back.

File: fibers.cpp (hunk 7 above)
Comment (leeeqian) [inline fibers.cpp]: Okay, I also added same verify when recycling from create()

**Comment by leeeqian** [comment]
currently yield() only switches the last unblocked fiber,
is it supposed to switch any ready fiber?

**Comment by apmckinlay** [comment]
Not sure I understand your question but I'll try to answer.
Every time yield is called from the main fiber, fi is incremented (possibly wrapping around), so each fiber should get a chance to run.
Unblocking is a special case. The fiber would have blocked before the end of its time slice, so it makes sense to let it run before its next "turn". Also, blocking is used for network activity and we want to handle that quickly. 
There is the potential for unfair scheduling if a fiber constantly blocked and unblocked. Other fibers might not get a chance to run, and the fiber after it would be more likely to run than the fiber before it.
But network activity is slow (relatively speaking) so other fibers should get a chance to run. And we don't run a lot of fibers at once, or rely on fair scheduling, so I don't think it's critical.

File: fibers.cpp [hunk 8]
```diff
@@ -238,20 +247,20 @@ bool Fibers::yield()
 		return true;
 		}
 
-	for (int i = 1; i < MAXFIBERS; ++i)
+	fi = fi % (MAXFIBERS - 1) + 1;
+	if (runnable(fibers[fi]))
```
Comment (apmckinlay) [inline fibers.cpp]: Why did this get moved outside the loop? A loop is necessary to search for a runnable fiber. 

Only cleaning up the ones we scan over is a way of limiting the cleanup work to keep the yield overhead low. It's not critical to clean up everything all at once. Either yield or create will eventually handle it.

Comment (leeeqian) [inline fibers.cpp]: I misunderstood the loop, I will put it back.

**Pull request merged**